### PR TITLE
Faster execution under `jax.vmap`

### PR DESCRIPTION
**Pull request by PhilipVinc**
**TLDR**: This PR improves the jax interface so that when `jax.vmap` is used to transform some code, the device batch-execution API is used (forward pass only, for now), as opposed to now that the batch-execution API is never used.

**Explanation**: In #3244 the jax interface was switched over to [`jax.pure_callback`](https://jax.readthedocs.io/en/jax-v0.3.23/_autosummary/jax.pure_callback.html) which allowed the interface to automatically work under `jax.vmap` transformations.

Unfortunately, by default `jax.pure_callback` assumes `vectorized=False`, meaning that it will call the callback multiple times instead of using the `batched` interface of some pennylane devices.
Quoting the docstring

> If vectorized is True, when the callback is mapped via jax.vmap, it will be called directly on inputs with leading batch dimensions instead of executing callback on each mapped input individually. The callback should also return outputs batched across the leading axis.

In some of my experiments with @StefanoBarison we noticed that this is far too slow in some situations, in particular when computing jacobians, so I would like to improve the situation by switching those callbacks to `vectorized=True`.

Unfortunately, as you can see from the docstring above, the convention used by jax is that the batch dimension should be the leading dimension, while pennylane assumes the opposite. 

This PR is a proof of concept of switching the batch dimension when under jax.vmap. (In the future, a more elegant solution should be possible using `jax.custom_vmap.***` but this is not really stable yet).
Ideally, i'd like to switch all callbacks in this file to `vectorized=True` but I have issues doing that for the gradient callback...

cc @antalszava

### CODE REVIEW

**Comment by josh146** [comment]
Thanks @PhilipVinc! Also ccing in (for interest) @rmoyard and @dwierichs :)

**Comment by PhilipVinc** [comment]
This could already be reviewed/merged, doing the same for the backward pass in a later PR. 
I've been trying to understand where exactly it breaks down but I don't have enough time this week....

**Comment by PhilipVinc** [comment]
What tests would you want to add?
This is already covered by tests unless I'm mistaken.

**Comment by rmoyard** [comment]
Oh I think I misunderstood, here we fix the vectorization for PennyLane broadcasting but it does not support `jax.vmap` yet?

**Comment by PhilipVinc** [comment]
Right now, `jax.vmap(qml.qnode(fun))(x)` gets traced by jax as 

```python
QML.qnode (fun)(x[0])
QML.qnode (fun)(x[1])
...

QML.qnode (fun)(x[N])
```
Which will never leverage the batch-execution interface some devices have.

With my PR, this gets traced as 

```python
QML.qnode (fun)(x)
```

Which can leverage the batch execution interface of some Pennylane devices.

It literally does not change the user-facing API. It's just an optimization in how Pennylane should be called from jax.

**Review by rmoyard — APPROVED** [review]
Thank you for the clarification!

**Comment by antalszava** [comment]
Hi @PhilipVinc, thank you for the addition here!

Two things come to mind:
1. Does turning on vectorization in the non-vectorized case have any performance implications?
2. The `jax.vmap(qml.qnode(fun))(x)` cases using the JAX-JIT interface are not explicitly tested in our test suite (but rather parameter broadcasting is), so it can be worth adding a couple of such cases at one point. What would be an example that could be good to test?

**Comment by PhilipVinc** [comment]
Another way to see what this PR does:

```python

In [1]: from jax.config import config
   ...: config.update("jax_enable_x64", True)
   ...:
   ...: import jax
   ...: import jax.numpy as jnp
   ...: import numpy as np
   ...: import pennylane as qml
   ...:
   ...: phys_qubits = 2
   ...: n_configs = 5
   ...: np.random.seed(1)
   ...: pars_q      = np.random.rand(n_configs,2)
   ...: pars_q1      = np.random.rand(2, n_configs)
   ...:
   ...: dev = qml.device("default.qubit", wires=tuple(range(phys_qubits)), shots=None)
   ...:
   ...: def minimal_circ(params):
   ...:     @qml.qnode(dev, interface="jax-jit",diff_method="parameter-shift", cache=None)
   ...:     def _measure_operator():
   ...:         qml.RY(params[...,0],wires=0)
   ...:         qml.RY(params[...,1],wires=1)
   ...:
   ...:         op = qml.Hamiltonian([1.0],[qml.PauliZ(0) @ qml.PauliZ(1)])
   ...:         return qml.expval(op)
   ...:     res = _measure_operator()
   ...:     return res
   ...:
```

standard Pennylane broadcasting it calls the callback with two arrays (`c` and `e`) of length 5 and returns a matrix of shape (5,1):
```python
{ lambda ; a:f64[5,2] # input pars_q
    c:f64[5] = ... # a[...,0]
    e:f64[5] = ... # a[...,1]
    f:f64[1,5] = custom_vjp_call_jaxpr[...
      fun_jaxpr={ lambda ; g:f64[5] h:f64[5]. let
          i:f64[1,5] = pure_callback[
            callback=<function pure_callback.<locals>._flat_callback at 0x111745e10>
            result_avals=(ShapedArray(float64[1,5]),)
            vectorized=False
          ] ...
    ] c e  # calls the callback with c and e that are two arrays of length 5. returns f which is a [1,5] matrix
```

current master (vectorised = False)
```python
{ lambda ; a:f64[5,2] # input pars_q
    c:f64[5] = ... # a[...,0]
    e:f64[5] = ... # a[...,1]
    f:f64[5,1] = custom_vjp_call_jaxpr[
      ...
      fun_jaxpr={ lambda ; g:f64[5] h:f64[5]. let
          i:f64[5,1] = scan[                                            # Notice here the scan, which is a for loop around the callback
            jaxpr={ lambda ; j:f64[] k:f64[]. let
                l:f64[1] = pure_callback[
                  callback=<function pure_callback.<locals>._flat_callback at 0x11179ed40>
                  result_avals=(ShapedArray(float64[1]),)                #  The output of the callback, each time it's called, is a single scalar
                  vectorized=False  # vectorized=False as in master 
                ] j k
              in (l,) }
            length=5 # length of the for loop
          ] g h
...
```

Finally, this PR: 
```python
In [2]: jax.make_jaxpr(minimal_circ)(pars_q)
    c:f64[5] = ... # a[...,0]
    e:f64[5] = ... # a[...,1]
    f:f64[5,1] = custom_vjp_call_jaxpr[...
      fun_jaxpr={ lambda ; g:f64[5] h:f64[5]. let
          i:f64[5,1] = pure_callback[                           # notice that there is no scan loop, so the callback is called only once, like in the non-vmap case.
            callback=<function pure_callback.<locals>._flat_callback at 0x13c849fc0>
            result_avals=(ShapedArray(float64[5,1]),)
            vectorized=True # note the vectorized true keyword
          ] g h
        in (i,) }...
    ] c e
```

admittedly reading this IR is not trivial, but from it you can see that essentially this PR makes it so that the `jax.vmap` code becomes identical to the non-hax.vmap code, while before there was a slowdown induced by the scan loop.

**Comment by PhilipVinc** [comment]
Thanks for pushing the test.

Moving this over the new interface should be a trivial copy-paste job. I can do it right now if you want.

**Comment by antalszava** [comment]
Oh @PhilipVinc,

> Don't forget to add it and yourself to the changelog.

The last step here is to:

> Write a changelog entry for all Pull Requests. The changelog entry should include a link back to the PR. More significant user-facing changes should include a code example. In addition to the changelog entry itself, add your name to the alphabetical contributors’ list at the bottom of each release’s section. The changelog is located at doc/releases/changelog-dev.md.

(https://docs.pennylane.ai/en/stable/development/guide/pullrequests.html#ready-for-review)

**Comment by antalszava** [comment]
> Moving this over the new interface should be a trivial copy-paste job. I can do it right now if you want.

Sure, thank you for offering! That would be great. We may get to it with some delay.

**Comment by PhilipVinc** [comment]
where should I enable the tests for the new interface?

**Comment by antalszava** [comment]
https://github.com/PennyLaneAI/pennylane/blob/master/tests/returntypes/test_jax_jit_qnode_new.py#L1609 This could be a good spot to add the test.

**Comment by PhilipVinc** [comment]
@antalszava I also added a test for multi-expval qnodes to make sure it does not break to both interfaces.

**Comment by antalszava** [comment]
@PhilipVinc thank you!
